subprojects/core/src/main/java/org/gradle/execution/plan/RelatedLocations.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.execution.plan;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import org.gradle.internal.snapshot.CaseSensitivity;
+import org.gradle.internal.snapshot.ChildMap;
+import org.gradle.internal.snapshot.ChildMapFactory;
+import org.gradle.internal.snapshot.EmptyChildMap;
+import org.gradle.internal.snapshot.VfsRelativePath;
+
+import java.util.Collection;
+import java.util.function.Consumer;
+
+public class RelatedLocations {
+    private volatile RelatedLocation root = new DefaultRelatedLocation(EmptyChildMap.getInstance(), ImmutableList.of());
+
+    public Collection<Node> getNodesRelatedTo(String location) {
+        VfsRelativePath relativePath = VfsRelativePath.of(location);
+        ImmutableSet.Builder<Node> builder = ImmutableSet.builder();
+        if (relativePath.length() == 0) {
+            root.visitNodes(builder::add);
+        } else {
+            root.visitNodes(relativePath, builder::add);
+        }
+        return builder.build();
+    }
+
+    public synchronized void recordRelatedToNode(Node node, Iterable<String> locations) {
+        for (String location : locations) {
+            VfsRelativePath relativePath = VfsRelativePath.of(location);
+            root = root.recordRelatedToNode(relativePath, node);
+        }
+    }
+
+    public synchronized void clear() {
+        root = new DefaultRelatedLocation(EmptyChildMap.getInstance(), ImmutableList.of());
+    }
+
+    private interface RelatedLocation {
+        Collection<Node> getNodes();
+        void visitNodes(VfsRelativePath relativePath, Consumer<Node> nodeConsumer);
+        void visitNodes(Consumer<Node> nodeConsumer);
+        RelatedLocation recordRelatedToNode(VfsRelativePath relativePath, Node node);
+
+        ChildMap<RelatedLocation> getChildren();
+    }
```

Comment:
I don't think we should have a private interface for a single private implementation.

Reply:
I removed the interface.